compiler/llvm/clangUtil.cpp

```diff
@@ -1302,6 +1302,15 @@ class CCodeGenConsumer final : public ASTConsumer {
     bool parseOnly;
     ASTContext* savedCtx;
 
+    bool shouldHandleDecl(Decl* d) {
```

Comment:
This function confuses me a little bit.  Previously we added declarations to all C functions and now we only add them if they have the cuda `__device__` attribute.  Why does this fix the underlying problem?

If we're not using the gpu locale model would't we want to maintain the old behavior (indiscriminately bringing in all decls?)

Reply:
> Why does this fix the underlying problem?

We add `__device__` decls when we are codegening for the GPU, and non-`__device__` decls when we are codegening for the CPU. This avoids the problem of trying to jam both device and non-device decls into our table (LVT) which doesn't handle overloaded functions at all. Instead, we just add the one appropriate for what we are currently compiling (GPU device or not).

> If we're not using the gpu locale model would't we want to maintain the old behavior (indiscriminately bringing in all decls?)

Yes I think it should have `return true` at the end

Reply:
> We add __device__ decls when we are codegening for the GPU, and non-__device__ decls when we are codegening for the CPU

When you say codegening for the GPU do you mean generating code when `CHPL_LOCALE_MODEL=gpu` or do you doing `codegenPartTwo` with `gCodegenGPU=true`.

If you mean the latter your explanation makes sense to me, but then but then we'd want to do something like `if (gCodegenGPU) { ... }` and not `if(localeUsesGPU())`, right?

Reply:
> When you say codegening for the GPU do you mean generating code when CHPL_LOCALE_MODEL=gpu or do you doing codegenPartTwo with gCodegenGPU=true.

I meant codegenPartTwo with gCodegenGPU=true.

> If you mean the latter your explanation makes sense to me, but then but then we'd want to do something like if (gCodegenGPU) { ... } and not if(localeUsesGPU()), right?

If localeUsesGPU the logic is activated. Then, we use `gCodegenGPU` to check it matches the function having `__device__` in `gCodegenGPU == d->hasAttr<CUDADeviceAttr>();` so that we get device code only on device compiles and host code only on host compiles.

Reply:
Yep, sorry for the confusing and wrong logic here. I refactored it out of a more in-place check. Wasn't too careful doing that. It is very likely the root cause of the `make check` problem.